### PR TITLE
Databuilder docs spike: trying out importing all databuilder docs from databuilder repo

### PR DESCRIPTION
**Pull request by rebkwok**
See https://github.com/opensafely-core/databuilder/issues/808

Trying out the [multirepo-plugin ](https://github.com/jdoiro3/mkdocs-multirepo-plugin) by pulling out all of the databuilder docs to the databuilder repo (`docs-spike-2` branch).  

Imported docs are built under `data-builder-imported/` URLs.  No databuilder docs have been removed from this repo yet, so the original docs are under `data-builder/`

Some notes:
1) We need a nav entry for any imported repo (so we can't have the hidden URLs we use atm without some additional work)
2) Snippets from the imported repo will just work, as long as we define the base paths for both the main repo and the imported repo in the mkdocs.yml
3) the table plugin for reading CSV files (for databuilder tutorial examples) will not just work, because the plugin will only look at  the path from the main repo; I forked the plugin so it allows looking in multiple data paths, just to verify there's a workaround.  As this is only used for the databuilder docs, it could potentially go away when we the aren't trying to support keeping the docs in two places in parallel.
4) snippets in the imported repo can't refer to paths in the main repo (or at least they can, but the paths check will fail when running the docs in databuilder).  There are 2 options here; one is to put a dummy file in the expected snippet location in  databulder; the main docs path will be checked and loaded first.  The other option is to have a flag an rewrite the snippet marker, which is what I've done here, in `hooks/parent_snippet.py`.  At the moment the only snippet it does this for is the glossary, so the databuilder docs can use the main docs' glossary.
5) Currently when the full site builds, the specs, contracts and backends are still loaded from the `public_docs.json` at the root of this repo.  I haven't done anything about that yet because if we go forwards with this approach it's likely that we'll change the way those docs are generated anyway.

### CODE REVIEW

File: .github/workflows/pages-deployment.yml
Comment (StevenMaude) [inline .github/workflows/pages-deployment.yml:29]: question: Could we limit this token's permission? For example:

```yaml
permissions:
  contents: read

jobs:
  …
```

***

I also note that Cloudflare's publish action [uses it for deployments](https://github.com/cloudflare/pages-action/blob/1aff065c9e5139731bb0e61b9796b800c22279a0/README.md#usage). So actually the configuration might be something like:

```yaml
permissions:
  contents: read
  deployments: write
```

***

That said, I'm not entirely sure why the token is needed at all for the multirepo plugin. The documentation for the multirepo plugin does [state that you must provide a token](https://github.com/jdoiro3/mkdocs-multirepo-plugin/blob/4f7820717fcf687576075ec14813ecc71de06535/README.md#github-actions). But the actual cloning script used by it seems — to my naive eyes — to fallback to attempting to [clone without a token, if none is provided](https://github.com/jdoiro3/mkdocs-multirepo-plugin/blob/4f7820717fcf687576075ec14813ecc71de06535/mkdocs_multirepo_plugin/scripts/sparse_clone.sh#L12-L19).

Then again, without a token you may encounter rate limits to the IP address in GitHub Actions runners, I think, which could cause random failures. So that might be an argument for using a token with limited permissions.

Comment (rebkwok) [reply .github/workflows/pages-deployment.yml:29]: It does need the token - the repo doesn't get cloned without it (I tried).  I think it may be specific to cloning in Github Actions - it doesn't seem to need a token or auth locally.
We can try limiting the token; do we do that in other repos (and should we)?

File: hooks/parent_snippets.py
Comment (StevenMaude) [inline hooks/parent_snippets.py:21]: thought: This is a good workaround :+1:

I guess this cannot easily work in the child repository.

Another alternative would be to make the snippets a submodule or subtree that is imported by both parent and child. But that's then a third repository of content to maintain. And not sure if it's worth the extra complication.

File: mkdocs.yml
Comment (StevenMaude) [inline mkdocs.yml:140]: suggestion: Maybe set this `cleanup` value to be specified as an [environment variable](https://www.mkdocs.org/user-guide/configuration/#environment-variables) with a default of `false`? And then setting it to `true` only for the GitHub Actions build.

> Note: Cleanup should be set to `false` when developing (i.e., when calling `mkdocs serve`). This will prevent importing repos multiple times with livereload.

https://github.com/jdoiro3/mkdocs-multirepo-plugin/blob/4f7820717fcf687576075ec14813ecc71de06535/README.md#setup

File: requirements.prod.in
Comment (StevenMaude) [inline requirements.prod.in:11]: suggestion: If we go ahead with using this, it would be good to move this to opensafely-core and/or submit a PR to upstream to add `data_paths` as an option.

File: .github/workflows/pages-deployment.yml
Comment (StevenMaude) [reply .github/workflows/pages-deployment.yml:29]: Oh, OK! That's strange, because it looks like from the code that it should work, but I only glanced at it. Maybe it's different in GitHub Actions specifically to avoid rate-limiting failures.

> We can try limiting the token; do we do that in other repos (and should we)?

To my knowledge, we don't have any organisational policy on this. So maybe that's worth asking elsewhere.

(My personal view is: yes, based on the principle of least privilege. I think it's generally worth considering what the GitHub Actions tokens can do, particularly when interacting with third-party actions. The defaults are [quite permissive](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token), unless they are already locked down by the organisation owners.)

File: hooks/parent_snippets.py
Comment (StevenMaude) [reply hooks/parent_snippets.py:21]: That all seems reasonable; I think you're correct that the glossary is the main use case right now :heavy_check_mark:

File: requirements.prod.in
Comment (StevenMaude) [reply requirements.prod.in:11]: Also fine to leave it until we decide if it's still needed upon moving out the Data Builder docs.

**Comment by StevenMaude** [comment]
This looks broadly fine to me overall, so I'll approve. I think we need to:

* rebuild the requirements file build to be for Python 3.9
* remove the Data Builder (imported) entry from the navigation

before this is merged. We also need to decide on how migration to this new system might work.

We discussed most of the caveats mentioned already in this thread, and the limitations seem relatively small, compared with the benefits of keeping the documentation closer to the code.

If you want to fully preview the documentation, then you should do a full documentation site build. But if you just want to see the changes your making to the Data Builder component's documentation, then it's sufficient to run a local build, with those caveats.

**Review by StevenMaude — APPROVED** [review]

**Comment by StevenMaude** [comment]
# Migration

There are two options:

1. Immediately replace the existing `docs/data-builder` content with this new build method.
2. Keep the Data Builder documentation duplicated for now, until we definitively move.

2 seems like it would be less invasive. It means that nothing in the "real" documentation should ever be broken. It does mean that we have some duplication of content that may later go out of sync with the canonical version.

What we could do to mitigate the duplication:

* Remove the extra "Data Builder (imported)" entry from the navigation.
  * This is desirable anyway because we don't want to make the Data Builder documentation too easily discoverable yet).
* Remove the extra "Data Builder (imported)" content from site search
  * This is so that users don't accidentally end up with this in their browser history/bookmarks, or get confused by duplicate pages. Particularly since the two "duplicate" entries actually go out of sync in the short-term.
  * Material for MkDocs does have search exclusion but seems you must specify this [per page](https://squidfunk.github.io/mkdocs-material/setup/setting-up-site-search/#search-exclusion), as far as I can tell.
  * There is a plugin we could consider using temporarily: https://github.com/chrieke/mkdocs-exclude-search — this allows for more blanket exclusions.
* Consider adding a robots.txt to ensure the additional "Data Builder (imported)" URLs doesn't end up in search engines.
  * Google is already crawling at least some of the Data Builder and ehrQL pages. The documentation site may get penalised for having duplicate content without a canonical link element. And duplicates may also confuse anyone searching via a search engine.

We'll also need to ensure that the documentation content in the Data Builder repository is entirely synchronised with that in the main documentation repository, before removing the Data Builder content from the documentation repository. In the meantime, any changes should be to the main documentation repository.

File: requirements.prod.in
Comment (rebkwok) [reply requirements.prod.in:11]: Sure - this was only temporary.  There's a bug for wider use anyway (it can deal with more than one `data_path`, but not if there's more than one used in the same file)

File: requirements.prod.txt
Comment (rebkwok) [reply requirements.prod.txt:2]: huh, must have been using the wrong venv

**Comment by rebkwok** [comment]
@StevenMaude Initially I was also thinking that migration option #2 was the way to go, but I've changed my mind. 

If we immediatedly replace the current docs:
1) There's no need to attempt to keep the documentation up to date in 2 places, which will just be a pain - something that anyone who makes a change needs to remember to fix up. The most likely thing is that a change gets made in the main docs and forgotten in the databuilder ones, and then when we do move over to the new way it gets lost.
2) We don't need the hacky fix for the tables plugin
3) We don't need any temporary workaround for searching
4) We don't actually lose any of the old docs and can always roll back
5) All the current users of the docs are internal, so if something isn't quite right, they can tell us.  And we already have the warning notices in there about databuilder not being fully ready yet.
 6) We can be sure that the new docs are building properly and not being affected by the old ones still in this repo.  E.g. I know that the docs built from the specs tests are only building here because of the `public_docs.json` that's here for the old docs.  Any snippet includes in the new docs *should* be using their own version of the snippet folders, but with the old docs still there, it's not easy to be sure. 

If we do this, the main thing we still need to do is to hide the nav entry (it needs to be in the nav or the multirepo plugin won't work, but it should be fairly easy to set a flag and hide it when the docs build).

File: .github/workflows/pages-deployment.yml
Comment (rebkwok) [reply .github/workflows/pages-deployment.yml:34]: Done and all looks fine in the preview

File: mkdocs.yml
Comment (rebkwok) [reply mkdocs.yml]: Done :)

File: requirements.prod.txt
Comment (rebkwok) [reply requirements.prod.txt]: Apparently I was using the right venv, but from pre-3.9.  Rebuilt and nothing changed, but I've updated the req file anyway

**Comment by StevenMaude** [comment]
@rebkwok: :+1: that sounds reasonable

We can reasonably easily immediately revert the change if we need to back it out. That's provided that we have an informal "mutex lock" on the documentation content for a brief period and consider it unlocked once we're happy with the move. In practice, it's been mostly myself editing those pages, so I think that's fine.

I don't think there's too much risk here; we've already seen it all working with the preview.

Also, I checked there aren't any (substantial) open PRs with changes to these files either. There is #985 but that's just adding one `"` character, and I don't think it would merge in anyway right now because the file got renamed. I can just fix that up after the move.

***

In that case, then I think the remaining tasks to get the initial move complete are:

1. Possibly do a final resync of the `docs/data-builder` content in the documentation repository to the associated [Data Builder PR](opensafely-core/databuilder#938). This is so that the PR contains the final state of content from the documentation repository. However, I'm not sure it's entirely necessary though: the last edit here was at 2023-01-12 17:17 [^1]. 
2. Merge that PR opensafely-core/databuilder#938.
3. Merge this PR.
4. Look over the docs and check for any immediate problems.
5. Make a small change to the docs in the Data Builder repository and ensure it filters through; the above single character change would be fine.
6. Consider when to remove the vestigial `docs/data-builder` content from here.

Please shout if there's anything I've overlooked, or you have other thoughts.

[^1]: Being easily able to see when the contents by checking the modification time of the entire directory is another nice consequence of pulling out the content into its own hierarchy.

**Comment by rebkwok** [comment]
I've added a couple more commits that deal with the CI config and hiding the nav section.

I think removing the docs/databuilder section is something we should do as part of this PR; it'll mean we can be sure that the docs being built are using the files we think they are.

The one other piece is making merges to main on databuilder trigger a rebuild of these docs - but that can wait until after these PRs are merged, I think.  

One thing to consider: databuilder updates on the secure server(s) are manual, as they are for cohort-extractor.  Triggering the docs update on merges means that the docs will reflect the state of the main branch immediately, and the version of databuilder on the server might lag behind.  Hopefully by a matter of hours at the most, since anyone merging a databuilder PR should be aware that they need to also pull the new image on the server (and this issue already exists for cohort-extractor anyway, so it's nothing new)

**Comment by rebkwok** [comment]
I think the order to do this is:
1) Do the final resync of files on the databuilder PR (If we think it's necessary)
2) Update this PR to:
    - remove docs/databuilder folder
    - remove the table plugin workaround since we won't need it anymore
    - point the imported repo at /data-builder instead of /data-builder-imported
3) Push and check the deploy preview still looks OK
4) Merge the databuilder PR
5) Update this PR to:
    - point multirepo URL to databuilder `main` instead of the PR branch
6) Merge this PR

**Comment by StevenMaude** [comment]
All of [this comment](https://github.com/opensafely/documentation/pull/1095#issuecomment-1396875562) sounds good to go ahead with :+1:

I agree that the "public docs slightly ahead of actual deployment for a few hours or a day" issue is fine as a known issue; and, as you say, it's been the same for cohort-extractor anyway.

**Comment by rebkwok** [comment]
The name of the nav section is what I meant by that comment :)

OK - I'll make a start on 1 & 2

**Comment by rebkwok** [comment]
An update on where we are now:

**This (main docs) repo**:
- all databuilder docs have been moved out of this repo (both the databuilder/ and docs/databuilder folders)
- public_docs.json has been removed, and replaced by a hook that copies the imported file to the expected location
- Added a hook to remove the nav section for the imported repo
- The imported repo's docs are now built at /data-builder/ URLS
- The check_snippets.yml workflow is removed, since the snippets aren't here anymore 

**Databuilder repo** (https://github.com/opensafely-core/databuilder/pull/938):
- All docs from here are moved to the root docs/ folder (updated, but nothing new to add in the docs content since the PR was opened)
- the public_docs.json is under the docs/ folder too, and removed from .gitignore (we may well change the way the contracts/backends/specs are built and do away with the need for this file, and mkdocs-opensafely-databuilder, but that's for another day)
- Moved the databuilder-related just commands to the databuilder justfile and updated them to work from within the local repo
- Added the workflow in check_snippets.yml to the databuilder repo.  It's currently only set to run on `workflow_dispatch` because the tutorial examples need fixing (one of them is calling a `subtract_days()` that doesn't exist), and I don't want to fix them now.
- Removed the `update_public_docs.yml` workflow from the databuilder repo, and added a new step in the docs checks workflow to generate a new public_docs.json and check it hasn't changed (for some reason that isn't working yet - it detects a tiny change and I'm not sure why; it may be irrelevant if we change the way those docs are built, so I'm not worrying too much about it now)

**Still to do before this PR is merged**
- [ ] Check the databuilder docs in the deploy preview
- [ ] Merge databuilder PR
- [ ] Update multirepo config to use `main`

Other things to remember to do later:
- [ ]  Add a step on databuilder to trigger re-deploy on this repo
- [ ] fix tutorial examples
- [ ] add workflow on databuilder to check links (note we'll need a workaround for links that refer to the main docs - they're relative links so they work when the databuilder docs are imported to the main build, but they won't work in the local repo)
- [ ] Update check_docs workflow on databuilder to run properly and on push/PRs
- [ ] Check updates to the databuilder docs make it through to the main docs (fixing the broken tutorial example could be a test for that)

**Comment by StevenMaude** [comment]
> Other things to remember to do later:
> …

I've moved all of these into a list on the [main migration issue](https://github.com/opensafely-core/databuilder/issues/808#issuecomment-1379192630)

The one I'm not sure about is "Fix tutorial examples", if there is more than the one in #985 to do, or something else that needs updating.

**Pull request merged**
